### DISCUSSION TITLE
Thoughts on keyboard access to the IDE

### DISCUSSION THREAD

**Discussion by gbishop** [gbishop/OS-DPI#10] [Ideas]
# Thinking about TreeBase access

Can we come up with an approach to keyboard access that meets the ARIA guidelines
and works for all our controls?

Here is an abbreviated example of a typical tree from the Method tab. This is part
of a list of methods.

```html
<fieldset class="Method">
  <legend>Method: My method</legend>
  <label>
    <span>Name</span>
    <input type="text" value="My method" />
  </label>
  <label>
    <span>Active</span>
    <input type="checkbox" checked="" />
  </label>

  <fieldset>
    <legend>
      Timers
      <button class="addChildButton" title="Add a timer">+</button>
    </legend>

    <ul>
      <li>
        <label hiddenlabel="">
          <span>Name</span>
          <input type="text" />
        </label>
        <label hiddenlabel="">
          <span>Interval</span>
          <input type="number" step="any" />
        </label>
      </li>
    </ul>
  </fieldset>
  <fieldset>
    <legend>
      Handlers
      <button class="addChildButton" title="Add a key handler">+Key</button>
      <button class="addChildButton" title="Add a pointer handler">
        +Pointer
      </button>
      <button class="addChildButton" title="Add a timer handler">+Timer</button>
    </legend>
    <ol>
      <li>
        <fieldset class="Handler">
          <legend>Key Handler</legend>

          <label>
            <span>Signal</span>
            <select required="">
              <option value="" disabled="">Choose one...</option>
            </select></label
          >
          <label>
            <span>Debounce</span>
            <input type="number" step="any" />
          </label>

          <fieldset class="Keys">
            <legend>
              Keys
              <button class="addChildButton" title="Add a key">+</button>
            </legend>
            <ul>
              <li>
                <div class="Key">
                  <label hiddenlabel="">
                    <span>Key</span>
                    <select required="">
                      <option value="" disabled="">Choose one...</option>
                    </select>
                  </label>
                </div>
              </li>
            </ul>
          </fieldset>
          <fieldset class="Conditions">
            <legend>
              Conditions
              <button class="addChildButton" title="Add a condition">+</button>
            </legend>
            <ul></ul>
          </fieldset>
          <fieldset class="Responses">
            <legend>
              Responses
              <button class="addChildButton" title="Add a response">+</button>
            </legend>
            <ul>
              <li>
                <div class="Response">
                  <label hiddenlabel="">
                    <span>Response</span>
                    <select required="">
                      <option value="" disabled="">Choose one...</option>
                      <option value="Responder">none</option>
                    </select></label
                  >
                </div>
              </li>
            </ul>
          </fieldset>
        </fieldset>
      </li>
    </ol>
  </fieldset>
</fieldset>
```

A Method has Name and Active properties and children that define Timers and Handlers.
Shown here is a Key Handler includes properties to describe the Event it handlers,
any Conditions that limit when it applies, and its Response to the Event.

From the Designer's view, Timers and Handlers are just complex Properties.

- In this example you'd focus on the first fieldset and hopefully hear
  something like "Method: the method name".
- Enter would take you inside the fieldset.
- Right arrow would take you to the labels on the Inputs and enter would let you edit them.
- More right arrows take you to the Timers, and Handlers fieldsets.
- Hit Enter to go into the Handlers, you are presented with a list of Handlers.
- Stepping through them you hear the (say) Signal each is handling. Enter lets you edit one.
- Inside the Handler you visit its Properties and Keys (if present), Conditions, and Responses sets.

I think this could be described as a State Machine where the State is your location in
the DOM (with any information you need about type, parents, etc) and the input is the KeyboardEvent.

**Comment by christineiym** [comment]
I checked out keyboard access to Excel 365 and tried to find other examples (but am still coming up mostly short; thinking more generally, the closest I can come up with to an accessible editor is the VSCode editor (which has menus, search shortcuts, and a document tree that could be analogous to our layout tree, but I will have to research it more). I did compile my notes [here](https://github.com/christineiym/OS-DPI/blob/access/thinking/KeyboardAccess.md#an-alternative-model-for-keyboard-access-in-the-editor) while proposing an alternative model for keyboard access in the editor. I think you touched on this briefly (and I elaborate more in my notes), but maybe the IDE as a whole can be thought of and navigated as a permanently expanded tree?

**Comment by christineiym** [comment]
While experimenting in the repository, I was thinking of the accessibility of the tree editor in the context of the larger IDE, so I did start experimentation by [starting to implement accessibility of the IDE tabs](https://github.com/gbishop/OS-DPI/commit/65662810ac2563b9f1e1a41253c7227dd507eee7) with guidelines and JS code I found online (this is still a work in progress and has most of the framework as described in the links [here](https://github.com/christineiym/OS-DPI/blob/access/thinking/KeyboardAccess.md#tab-access-standards), but does not yet work. However, depending on how we want to think of the IDE as a whole (as a tree, or should the tabs be thought of as separate from their tree-like content?), my implementation might change.

**Comment by christineiym** [comment]
Here's a working file enumerating the IDE's capabilities: https://github.com/christineiym/OS-DPI/blob/main/thinking/interface-capabilities.md
